File: 106-construct-binary-tree-from-inorder-and-postorder-traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder == null || postorder == null || inorder.length != postorder.length) return null;
        HashMap<Integer, Integer> hm = new HashMap<Integer,Integer>();
        for (int i = 0; i < inorder.length; ++i) hm.put(inorder[i], i);
        return buildTreePostIn(0, inorder.length-1, postorder, 0, postorder.length-1,hm);
    }

    private TreeNode buildTreePostIn(int is, int ie, int[] postorder, int ps, int pe, HashMap<Integer,Integer> hm){
        if (ps > pe || is > ie) return null;
        TreeNode root = new TreeNode(postorder[pe]);
        int rV = hm.get(postorder[pe]);
        root.left =  buildTreePostIn(is, rV-1, postorder, ps, ps+rV-is-1, hm);;
        root.right = buildTreePostIn(rV+1, ie, postorder, ps+rV-is, pe-1, hm);;
        return root;
    }
}

File: 106-construct-binary-tree-from-inorder-and-postorder-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        index_map = {k: idx for idx, k in enumerate(inorder)}
        
        def build(left, right):
            if left >= right:
                return None
            
            root = TreeNode(postorder.pop(-1))
            root.right = build(index_map[root.val]+1, right)
            root.left = build(left, index_map[root.val])
            return root
        
        return build(0, len(inorder))
